### ISSUE TITLE
LambdaLR + class binding causes GPU memory leakage

### ISSUE THREAD

**Issue by kaniblu** [pytorch/pytorch#47589]
## 🐛 Bug

When `torch.optim.lr_scheduler.LambdaLR` is used in a specific way (where the optimizer and the scheduler are both attached to an object instance and the lambda function refers to an instance variable), a GPU memory leak is observed.

## To Reproduce

Run the following code to reproduce the behavior.

```python
import torch


weight = torch.nn.Linear(100, 100).to("cuda")


class Foo:
    def __init__(self):
        self.lr = 1.0
        self.op = torch.optim.Adam(weight.parameters())
        self.sc = torch.optim.lr_scheduler.LambdaLR(self.op, lr_lambda=lambda s: self.lr)

    def run(self):
        loss = weight(torch.randn(32, 100).to("cuda")).mean()
        self.op.zero_grad()
        loss.backward()
        self.sc.step()
        self.op.step()


while True:
    Foo().run()
```

A gradual increase in peak memory usage is observed in the main GPU device. This will cause OOM error eventually.

**Note**: The reason why the LR scheduler is allocated for each loop is I need to run the same code multiple times in a single process. The above snippet is just a barebone structure of the problematic code.

## Expected behavior

The memory leak should not happen. For contrasting examples, the following code examples do not exhibit memory leaks.

1. The LR scheduler does not make a reference to any instance-specific variables.

```python
import torch


weight = torch.nn.Linear(100, 100).to("cuda")


class Foo:
    def __init__(self):
        self.lr = 1.0
        self.op = torch.optim.Adam(weight.parameters())
        self.sc = torch.optim.lr_scheduler.LambdaLR(self.op, lr_lambda=lambda s: 1.0)  # no reference to the instance field

    def run(self):
        loss = weight(torch.randn(32, 100).to("cuda")).mean()
        self.op.zero_grad()
        loss.backward()
        self.sc.step()
        self.op.step()


while True:
    Foo().run()
```

2. No OOP involved.

```python
import torch


weight = torch.nn.Linear(100, 100).to("cuda")


while True:
    op = torch.optim.Adam(weight.parameters())
    sc = torch.optim.lr_scheduler.LambdaLR(op, lr_lambda=lambda s: 1.0)
    loss = weight(torch.randn(32, 100).to("cuda")).mean()
    op.zero_grad()
    loss.backward()
    sc.step()
    op.step()
```

3. Also note that if the LR scheduler is initialized once only, the subsequent code (`loss.backward()` and `op.step()`) will not cause the memory leak.

## Environment

```
PyTorch version: 1.8.0.dev20201021+cu110
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 3090
Nvidia driver version: 455.23.04
cuDNN version: /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.3
[pip3] torch==1.8.0.dev20201021+cu110
[pip3] torchfile==0.1.0
[pip3] torchtext==0.6.0
[conda] numpy                     1.18.3                   pypi_0    pypi
[conda] torch                     1.8.0.dev20201021+cu110          pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchtext                 0.6.0                    pypi_0    pypi
```

 - Any other relevant information:

## Additional context

Note that for my application, allocating a new LR scheduler for each loop is absolutely necessary. My current SW design is OOP-oriented, so I would really like this to be fixed soon. At the moment, I will probably refrain from using LambdaLR but IMO this is a critical bug that can affect all production use cases.

cc @ezyang @gchanan @zou3519 @bdhirsh @vincentqb

**Comment by heitorschueroff** [comment]
Making this high priority because it's causing OOM error.

**Comment by gchanan** [comment]
well it's causing an OOM error because it's leaking memory and running in a loop forever.

**Comment by malfet** [comment]
Most likely adding `gc.collect()` to the loop body will fix the issue, wouldn't it?

**Comment by albanD** [comment]
I would agree with @malfet : Python gc cannot detect the memory increase as it happens on the GPU and so will never run. So the cycle will never be collected unless you explicitly run the gc.

**Comment by albanD** [comment]
cc @colesbury do you know if there is a way to better advertise Tensor's memory usage (both CPU and GPU) to the python GC to try and reduce the impact of such issue?

**Comment by colesbury** [comment]
Don't create cyclic trash involving tensors?

Part of the problem is that we've added an API that makes creating cycles too easy (via lr_lambda). An alternative API in which you set the learning rate imperatively would not have had this problem (e.g. `scheduler.lr = 0.01`). We should avoid this pattern in future APIs.

You can try to call the GC on allocation failure, but I'm not a big fan of that strategy for a few reasons:
1) Risk of deadlock when acquiring the GIL from arbitrary C++ code
2) Risk of grinding to a halt when at or near memory limit
3) Linux usually runs with overcommit, so you might never have an allocation failure; just get killed by the OOM killer

**Comment by albanD** [comment]
I do agree that trying to collect on allocation failure is going to introduce more issues.

Closing this as user error: You should not create python object cycles in a tight loop. And if you do so, you might have to call explicitly python's gc if it doesn't run enough.
